### PR TITLE
Introduce unit test for checkout.

### PR DESCRIPTION
**Pull request by godrei**
### Checklist

- [x] I've read and accepted the [Contribution Guidelines](https://github.com/bitrise-steplib/.github/blob/main/CONTRIBUTING.md)
- Requires _MINOR_ [version update](https://semver.org/)

### Context

In this PR I plan to introduce unit testing for the checkout functionality of the step.

### Changes

- Introduce `CmdRunner` interface for `Output`, `Run`, and `RunWithRetry` functions and move the current implementation to `DefaultRunner`, this allows us to mock the runner in unit tests.
- Introduce `MockRunner` which does not execute commands, but stores their PrintableCommandArgs()
- Simplify `runWithRetry`
- Move checkout functionality into `checkoutState`, so we can set up isolated unit tests

### CODE REVIEW

File: gitclone/git.go [hunk 1]
```diff
@@ -115,46 +174,15 @@ func getDiffFile(buildURL, apiToken string, prID int) (string, error) {
...
-	err := c.SetStdout(os.Stdout).SetStderr(io.MultiWriter(os.Stderr, &buffer)).Run()
-	if err != nil {
-		if errorutil.IsExitStatusError(err) {
-			return errors.New(strings.TrimSpace(buffer.String()))
-		}
-		return err
-	}
-
-	return nil
+	return runner.Run(c)
```
Comment (adborbas) [inline gitclone/git.go:177]: What's the point of extracting this single line to a function?

Comment (adborbas) [reply gitclone/git.go:177]: Similar thing in line 181 and 185.

File: gitclone/gitclone_test.go [hunk 2]
```diff
@@ -0,0 +1,411 @@
...
+
+const always = 9999
+const rawCmdError = "dummy_cmd_error"
+
+func Test_checkoutState(t *testing.T) {
+	tests := []struct {
+		name       string
+		cfg        Config
+		cmdOutputs map[string]commandOutput
+		want       *step.Error
```
Comment (adborbas) [inline gitclone/gitclone_test.go:22]: `wantErr` would more reflect that this is the expected error.

File: gitclone/gitclone_test.go [hunk 3]
```diff
@@ -0,0 +1,411 @@
...
+		wantCmds   []string
+	}{
+		{
+			name: "Checkout commit",
+			cfg: Config{
+				RepositoryURL: "https://github.com/bitrise-io/steps-git-clone.git",
```
Comment (adborbas) [inline gitclone/gitclone_test.go:28]: Could you move the url and the commit to a constant?

File: gitclone/gitclone_test.go [hunk 4]
```diff
@@ -0,0 +1,411 @@
...
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			mockRunner := newMockRunner(tt.cmdOutputs)
+			runner = mockRunner
+			got := checkoutState(git.Git{}, tt.cfg)
+			if !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("checkoutState().err = (%#v), want %#v", got, tt.want)
+			}
+			// require.Equal(t, tt.want, got)
```
Comment (adborbas) [inline gitclone/gitclone_test.go:344]: Left a comment here.

File: gitclone/gitclone_test.go [hunk 5]
```diff
@@ -0,0 +1,411 @@
...
+			},
+		},
+		{
+			name: "Checkout nonexistent branch",
+			cfg: Config{
+				RepositoryURL: "https://github.com/bitrise-io/steps-git-clone.git",
+				Branch:        "fake",
+			},
+			cmdOutputs: map[string]commandOutput{
+				`git "fetch" "origin" "refs/heads/fake"`: {failForCalls: always},
```
Comment (adborbas) [inline gitclone/gitclone_test.go:87]: You could also use a [mocking package](https://github.com/stretchr/testify/#mock-package) to stub the mock runner to return various outputs for various inputs. Example for setting up this particular mocking package: https://github.com/bitrise-steplib/steps-generic-file-storage/blob/90872ec8e32248f5bbc4ce695e86c4db398ff269/bulkfiledownloader/mockFileDownloader_test.go#L25

File: gitclone/gitclone_test.go [hunk 6]
```diff
@@ -0,0 +1,411 @@
...
+
+	"github.com/bitrise-io/bitrise-init/step"
+	"github.com/bitrise-io/go-utils/command"
+	"github.com/bitrise-io/go-utils/command/git"
+)
+
+const always = 9999
+const rawCmdError = "dummy_cmd_error"
+
+func Test_checkoutState(t *testing.T) {
```
Comment (adborbas) [inline gitclone/gitclone_test.go:17]: Although I see the benefit of using table driven test for this scenario this test is more than 300 lines which makes it hard to understand. Could we group the scenarios in smaller functions?

File: gitclone/gitclone_test.go [hunk 7]
```diff
@@ -0,0 +1,411 @@
...
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			mockRunner := newMockRunner(tt.cmdOutputs)
+			runner = mockRunner
+			got := checkoutState(git.Git{}, tt.cfg)
+			if !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("checkoutState().err = (%#v), want %#v", got, tt.want)
+			}
+			// require.Equal(t, tt.want, got)
+			if !reflect.DeepEqual(mockRunner.Cmds(), tt.wantCmds) {
```
Comment (adborbas) [inline gitclone/gitclone_test.go:345]: I'm not sure what `reflect.DeepEqual` does but if just want to assert that two slices contains the same elements in the same order we could use the [testify/assert package](https://github.com/stretchr/testify/issues/759).

**Review by adborbas — CHANGES_REQUESTED** [review]

File: gitclone/git.go [hunk 8]
```diff
@@ -28,6 +28,65 @@ const (
...
+var runner CmdRunner = DefaultRunner{}
+
+// CmdRunner ...
+type CmdRunner interface {
+	Output(c *command.Model) (string, error)
+	Run(c *command.Model) error
+	RunWithRetry(c *command.Model) error
+}
+
+// DefaultRunner ...
```
[the remark this answers is not in the document]
Comment (lpusok) [reply gitclone/git.go]: Makes sense, moved.

File: gitclone/git.go [hunk 9]
```diff
@@ -28,6 +28,65 @@ const (
 	fetchFailedTag    = "fetch_failed"
 )
 
+var runner CmdRunner = DefaultRunner{}
+
+// CmdRunner ...
+type CmdRunner interface {
+	Output(c *command.Model) (string, error)
```
[the remark this answers is not in the document]
Comment (lpusok) [reply gitclone/git.go]: Renamed.

File: gitclone/git.go [hunk 10]
```diff
@@ -28,6 +28,65 @@ const (
 	fetchFailedTag    = "fetch_failed"
 )
 
+var runner CmdRunner = DefaultRunner{}
+
+// CmdRunner ...
+type CmdRunner interface {
```
[the remark this answers is not in the document]
Comment (lpusok) [reply gitclone/git.go]: Renamed.

File: gitclone/git.go (hunk 1 above)
Comment (lpusok) [reply gitclone/git.go]: Removed extra indirection.

File: gitclone/gitclone_test.go (hunk 2 above)
Comment (lpusok) [reply gitclone/gitclone_test.go]: Usually wantErr bool, but in this case makes sense, renamed.

File: gitclone/gitclone_test.go (hunk 3 above)
Comment (lpusok) [reply gitclone/gitclone_test.go]: I think excessive use of constants makes the tests harder to read.

File: gitclone/gitclone_test.go (hunk 6 above)
Comment (lpusok) [reply gitclone/gitclone_test.go]: Moved test cases out of the function, categorized them better.

File: gitclone/gitclone_test.go (hunk 7 above)
Comment (lpusok) [reply gitclone/gitclone_test.go:401]: It is doing the exact same thing, except require.Equal prints out a better diff on case of failure. Do not want to introduce the testify dependency for a single test yet.

File: gitclone/gitclone_test.go (hunk 3 above)
Comment (adborbas) [reply gitclone/gitclone_test.go]: I would agree with you if the actual value of repoURL and commit had an impact on the test outcome, but this is not the case, at least not in every test. By extracting them to constants you actually make the test less cluttered by only exposing values that are important for that particular test case.

File: gitclone/gitclone_test.go (hunk 5 above)
Comment (adborbas) [reply gitclone/gitclone_test.go]: Do you agree to this?

File: gitclone/gitclone_test.go (hunk 3 above)
Comment (lpusok) [reply gitclone/gitclone_test.go]: removed repo URL where not needed.

File: gitclone/gitclone_test.go (hunk 5 above)
Comment (adborbas) [reply gitclone/gitclone_test.go]: I'll create a new PR showcasing how that would look like.

**Review by adborbas — APPROVED** [review]

**Pull request merged**